apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/dynamic/ConfigurationDiscoveryService.java

```diff
@@ -238,17 +247,25 @@ private void getAgentDynamicConfig() {
      * Local dynamic configuration center.
      */
     public static class Register {
-        private final Map<String, WatcherHolder> register = new HashMap<>();
+        private final Map<String, List<WatcherHolder>> register = new HashMap<>();
 
         private boolean containsKey(String key) {
             return register.containsKey(key);
         }
 
         private void put(String key, WatcherHolder holder) {
-            register.put(key, holder);
+            List<WatcherHolder> watcherHolderList = register.get(key);
```

Comment:
As the watcher list for one key is supported, I think we have to add a sync to this method.

Reply:
> As the watcher list for one key is supported, I think we have to add a sync to this method.

I thought this sync problem, but if add syn,  `registerAgentConfigChangeWatcher` is synchronized, and `JDBCPreparedStatementSetterInterceptor` call the `registerAgentConfigChangeWatcher` method register the watcher  every time when prepared sql occur, I worry about it will effect the sql performance. And if we do not and syn to it, even if it register multi same wachers, it will not effect the `JDBCPreparedStatementSetterInterceptor` register case. Can we just add comment on `registerAgentConfigChangeWatcher` to announce it is thread unsafe?

Reply:
Then that is the issue of `JDBCPreparedStatementSetterInterceptor`. You should not register the watcher repeated, no matter with or without `sync`, it causes a huge perf. impact.

Reply:
Watcher should only be registered in the JDBC Driver initializing stage. Consider `JDBCDriverInterceptor` to initial this watcher.

Reply:
> Then that is the issue of `JDBCPreparedStatementSetterInterceptor`. You should not register the watcher repeated, no matter with or without `sync`, it causes a huge perf. impact.

Can I use static code block or constructor code block in `JDBCPreparedStatementSetterInterceptor` to register the watcher once?

Reply:
> Can we just add comment on registerAgentConfigChangeWatcher to announce it is thread unsafe?

It is pointless, you are using this at this concurrency status.  Do you.

Reply:
> Can I use static code block or constructor code block in JDBCPreparedStatementSetterInterceptor to register the watcher once?

Why do you have to do that? That is really bad for the agent in practice. Potentially, this could cause a lot of class loader issues.

Reply:
> > Can we just add comment on registerAgentConfigChangeWatcher to announce it is thread unsafe?
> 
> It is pointless, you are using this at this concurrency status. Do you.

I mean I add sync on `registerAgentConfigChangeWatcher` and remove the register logic out of `beforeMethod` in `JDBCPreparedStatementSetterInterceptor` using static or constructor code block. So it will not generate perf impact.

Reply:
`JDBCDriverInterceptor` is there, what do you concern?

Reply:
> > > Can we just add comment on registerAgentConfigChangeWatcher to announce it is thread unsafe?
> > 
> > 
> > It is pointless, you are using this at this concurrency status. Do you.
> 
> I mean I add sync on `registerAgentConfigChangeWatcher` and remove the register logic out of `beforeMethod` in `JDBCPreparedStatementSetterInterceptor` using static or constructor code block. So it will not generate perf impact.

I have answered this question.

**That is really bad for the agent in practice. Potentially, this could cause a lot of class loader issues.**

Reply:
`static` is never recommended in the interceptor, and all static blocks in the agent would be reviewed carefully. It is acceptable when it really has to be, and 100% correct. In the interceptor, I could jump to the conclusion, that `static` is not allowed.

Reply:
Ok, I have removed the register logic to `JDBCDriverInterceptor` and add `synchronized` on `registerAgentConfigChangeWatcher` method.